rust/rope/src/compare.rs

```diff
@@ -0,0 +1,550 @@
+    None
+}
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+pub fn ne_idx(one: &[u8], two: &[u8]) -> Option<usize> {
+    ne_idx_fallback(one, two)
+}
+
+#[inline]
+#[allow(dead_code)]
+#[doc(hidden)]
+pub fn ne_idx_fallback(one: &[u8], two: &[u8]) -> Option<usize> {
+    for i in 0..one.len().min(two.len()) {
+        if one[i] != two[i] { return Some(i); }
+    }
+    None
+}
+
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+pub fn ne_idx_rev(one: &[u8], two: &[u8]) -> Option<usize> {
+    ne_idx_rev_fallback(one, two)
+}
+
+#[inline]
+#[allow(dead_code)]
+#[doc(hidden)]
+pub fn ne_idx_rev_fallback(one: &[u8], two: &[u8]) -> Option<usize> {
+    let min_len =  one.len().min(two.len());
+    for i in 1..min_len {
+        if one[one.len()-i] != two[two.len()-i] { return Some(i - 1); }
+    }
+    None
+}
+
+/// Utility for efficiently comparing two ropes.
+pub struct RopeScanner<'a> {
+    base: Cursor<'a, RopeInfo>,
+    target: Cursor<'a, RopeInfo>,
+    base_chunk: &'a str,
+    target_chunk: &'a str,
+    scanned: usize,
+}
+
+impl<'a> RopeScanner<'a> {
+    pub fn new(base: &'a Rope, target: &'a Rope) -> Self {
+        RopeScanner {
+            base: Cursor::new(base, 0),
+            target: Cursor::new(target, 0),
+            base_chunk: "",
+            target_chunk: "",
+            scanned: 0,
+        }
+    }
+
+    /// Returns the distance to the left of the two offsets to the first codepoint
+    /// which is not equal in the two ropes. If no such position exists,
+    /// returns the distance to the nearest starting offset.
+    ///
+    /// if `stop` is not None, the scan will stop at if it reaches this value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use xi_rope::compare::RopeScanner;
+    /// # use xi_rope::Rope;
+    ///
+    /// let one = Rope::from("hiii");
+    /// let two = Rope::from("siii");
+    /// let mut scanner = RopeScanner::new(&one, &two);
+    /// assert_eq!(scanner.find_ne_char_left(one.len(), two.len(), None), 3);
+    /// assert_eq!(scanner.find_ne_char_left(one.len(), two.len(), 2), 2);
+    /// ```
+    pub fn find_ne_char_left<T>(&mut self, base_off: usize, targ_off: usize, stop: T) -> usize
+        where T: Into<Option<usize>>
+    {
+        let stop = stop.into().unwrap_or(usize::max_value());
+        self.base.set(base_off);
+        self.target.set(targ_off);
+        self.scanned = 0;
+
+        let (base_leaf, base_leaf_off) = self.base.get_leaf().unwrap();
+        let (target_leaf, target_leaf_off) = self.target.get_leaf().unwrap();
+
+        debug_assert!(self.target.is_boundary::<BaseMetric>());
+        debug_assert!(self.base.is_boundary::<BaseMetric>());
+        debug_assert!(base_leaf.is_char_boundary(base_leaf_off));
+        debug_assert!(target_leaf.is_char_boundary(target_leaf_off));
+
+        self.base_chunk = &base_leaf[..base_leaf_off];
+        self.target_chunk = &target_leaf[..target_leaf_off];
+
+        loop {
```

Comment:
Another way to write this kind of loop is to have a zip-like iterator over two ropes; that lets you separate out the chunk iteration logic (and reduces the amount of repeated code). But I think this is ok.